src/endian/detail/stream.hpp

```diff
@@ -16,26 +16,22 @@ namespace endian
 namespace detail
 {
 
+using const_stream = std::true_type;
```

Comment:
Maybe just do:

```
using const_stream = const uint8_t*;
using non_const_stream = uint8_t*;
```

Reply:
Sure, it would make the types much simpler if we then also removed the static asserts, or else I prefer this solution as both better convey the binary option and also keep the user from using the stream in a wrong way. So the question then is:
- Do we want to keep the types and code simpler?
- Or, do we want to save the developer from making a mistake?

:man_shrugging:

Reply:
I think I would prefer your suggestion, Morten.

As I move further and further away from making OOP API's I realize that the saying "WIth great power comes great responsibility" is very true when creating easily understandable, maintainable, and extendable code. Saving the user from shooting itself in the foot is for sure a nice thing, but from experience, I have come to find that keeping maintaining such code can be a bit of a counterweight to extend it, and when adding new options to an API often makes the guards a bit complicated and therefor the code less understandable. Why I mentioned that it was moving away from OOP that made me think of this, is because the OOP paradigm tends to favor easy and pretty usability over extendability, making the API very simple, but then if one wanted to extend things you need to create new classes, extend old classes, override member functions, hiding important logic away in constructors and so on. I've found that such scenarios are much easier to handle with single-responsibility free-functions with fitting names because what is happening at runtime is most of the time just right in front of you when reading the code, and if you want to change the behavior you don't need to make a new class and refactor your codebase to use that instead.

Thank you for listening to my rambling Ted talk

Reply:
To be honest, I don't really care :nerd_face:
One thing to note here is that the stream class is not intended to be used outside of the endian library - only stream_reader and writer are in the public (i.e., non detail namespace) API.
If the developer makes a mistake in stream he would do so in the endian library, and as the library is "guarded" by the unit tests this mistake should not leak out.

If you really want me to have an opinion, it would be to go with the simpler approach.

Reply:
> If you really want me to have an opinion, it would be to go with the simpler approach.

All right. Thank you. I assume you mean the simpler implementation rather than simpler usage/integration :smile:

Reply:
I think the simpler implementation is often to be favored - as it typically also easier to maintain, extend and even use. I don't mind objects etc. and I think it is a great mental model for designing code. 

Looking at it I dont' think the code in endian is a great example of OOP and the use of templates here makes the code really hard to understand/read. 

Over the years I've come to favor: "more but simple code" vs. "less but complicated code". 

If you can have less code and still keep it simple that's always a win - in this specific case I think the using statement just passing the `const uint8_t'* ` or `uint8_t*` is less code and simpler than the alternative.

Reply:
> in this specific case I think the using statement just passing the const uint8_t'* or uint8_t* is less code and simpler than the alternative

I do think that the use of constants and enums is important for making the code understandable. I would say that the benefits really outweigh the costs.